### PR TITLE
Fix broken referenceKey prop on SelectColumnsButton

### LINKED ISSUE #8431: DataGridConfigurable & SelectColumnsButton preferenceKey broken
**What you were expecting:**
When using the `<DataGridConfigurable>` component and apply a property `preferenceKey` with any value. You expect that a `<SelectColumnsButton>` component with the same property and value to show the columns of that datagrid.

**What happened instead:**
Instead we see an empty list of selectable columns.

**Related code:**
Following sandbox illustrates the problem on the Posts list page:
https://codesandbox.io/s/cocky-hooks-078gbb?file=/src/posts/PostList.tsx:2183-2196

**Environment**

* React-admin version: latest 
* React version: 18
* Browser: chrome

### PR DESCRIPTION
**Pull request by wcoppens**
Fixes #8431

This commit fixes the broken referenceKey prop for the SelectColumnsButton.

### CODE REVIEW

**Comment by WiXSL** [comment]
Test are failing

**Review by WiXSL — CHANGES_REQUESTED** [review]
Test are failing

File: packages/ra-ui-materialui/src/list/datagrid/SelectColumnsButton.tsx [hunk 1]
```diff
@@ -41,7 +41,7 @@ import { styled } from '@mui/material/styles';
 export const SelectColumnsButton = props => {
     const resource = useResourceContext(props);
     const preferenceKey =
-        props.preferenceKey || `preferences.${resource}.datagrid`;
+        `preferences.${props.preferenceKey}` || `preferences.${resource}.datagrid`;
```
Comment (WiXSL) [inline packages/ra-ui-materialui/src/list/datagrid/SelectColumnsButton.tsx:44]: Fix lint warning

**Review by WiXSL — CHANGES_REQUESTED** [review]

**Comment by wcoppens** [comment]
@WiXSL My bad, updated the codestyle and added some extra tests to verify working.

File: packages/ra-ui-materialui/src/list/datagrid/SelectColumnsButton.spec.tsx [hunk 2]
```diff
@@ -15,3 +15,15 @@ describe('<SelectColumnsButton>', () => {
         expect(screen.queryByText('1869')).not.toBeNull();
     });
 });
+
+describe('<SelectColumnsButton>', () => {
+    it('should render a datagrid with columns using the given preference key', async () => {
```
Comment (WiXSL) [inline packages/ra-ui-materialui/src/list/datagrid/SelectColumnsButton.spec.tsx:20]: Move this test inside the already defined describe function above

**Review by WiXSL — CHANGES_REQUESTED** [review]

**Review by WiXSL — APPROVED** [review]

File: packages/ra-ui-materialui/src/list/datagrid/SelectColumnsButton.tsx [hunk 3]
```diff
@@ -40,8 +40,9 @@ import { styled } from '@mui/material/styles';
  */
 export const SelectColumnsButton = props => {
     const resource = useResourceContext(props);
-    const preferenceKey =
-        props.preferenceKey || `preferences.${resource}.datagrid`;
+    const preferenceKey = props.preferenceKey
+        ? `preferences.${props.preferenceKey}`
+        : `preferences.${resource}.datagrid`;
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/list/datagrid/SelectColumnsButton.tsx:45]: For consistency, please use the same syntax as in `DatagridConfigurable`:

```jsx
    const finalPreferenceKey = preferenceKey || `${resource}.datagrid`;

    const [availableColumns, setAvailableColumns] = useStore<
        ConfigurableDatagridColumn[]
    >(`preferences.${finalPreferenceKey}.availableColumns`, []);

    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const [_, setOmit] = useStore<string[]>(
        `preferences.${finalPreferenceKey}.omit`,
        omit
    );
```

**Review by fzaninotto — CHANGES_REQUESTED** [review]

**Comment by fzaninotto** [comment]
Sorry, needs rebase

**Review by slax57 — CHANGES_REQUESTED** [review]
Could you please also apply the changes regarding the `props` type that are suggested in #8479
- typing `props` with `SelectColumnsButtonProps` 
- using the `sanitizeRestProps` function

Also, there are 2 linter warnings you need to fix:


Thanks

**Comment by septentrion-730n** [comment]
Hi @wcoppens 

I did not see your PR, sorry about that!

**Review by slax57 — APPROVED** [review]
Thanks!

**Pull request merged**